client/rest/src/main/java/org/elasticsearch/client/RestClient.java

```diff
@@ -102,53 +106,93 @@
+               List<Node> nodes, String pathPrefix, FailureListener failureListener) {
         this.client = client;
         this.maxRetryTimeoutMillis = maxRetryTimeoutMillis;
         this.defaultHeaders = Collections.unmodifiableList(Arrays.asList(defaultHeaders));
         this.failureListener = failureListener;
         this.pathPrefix = pathPrefix;
-        setHosts(hosts);
+        setNodes(nodes);
     }
 
     /**
      * Returns a new {@link RestClientBuilder} to help with {@link RestClient} creation.
      * Creates a new builder instance and sets the hosts that the client will send requests to.
+     * <p>
+     * Prefer this to {@link #builder(HttpHost...)} if you have metadata up front about the nodes.
+     * If you don't either one is fine.
+     */
+    public static RestClientBuilder builder(Node... nodes) {
+        return new RestClientBuilder(nodes == null ? null : Arrays.asList(nodes));
+    }
+
+    /**
+     * Returns a new {@link RestClientBuilder} to help with {@link RestClient} creation.
+     * Creates a new builder instance and sets the nodes that the client will send requests to.
+     * <p>
+     * You can use this if you do not have metadata up front about the nodes. If you do, prefer
+     * {@link #builder(Node...)}.
+     * @see Node#Node(HttpHost)
      */
     public static RestClientBuilder builder(HttpHost... hosts) {
-        return new RestClientBuilder(hosts);
+        return new RestClientBuilder(hostsToNodes(hosts));
     }
 
     /**
-     * Replaces the hosts that the client communicates with.
-     * @see HttpHost
+     * Replaces the hosts with which the client communicates.
+     *
+     * @deprecated prefer {@link setNodes} because it allows you
+     * to set metadata for use with {@link NodeSelector}s
      */
-    public synchronized void setHosts(HttpHost... hosts) {
-        if (hosts == null || hosts.length == 0) {
-            throw new IllegalArgumentException("hosts must not be null nor empty");
+    @Deprecated
+    public void setHosts(HttpHost... hosts) {
+        setNodes(hostsToNodes(hosts));
+    }
+
+    /**
+     * Replaces the nodes with which the client communicates.
+     */
+    public synchronized void setNodes(Collection<Node> nodes) {
+        if (nodes == null || nodes.isEmpty()) {
+            throw new IllegalArgumentException("nodes must not be null or empty");
         }
-        Set<HttpHost> httpHosts = new LinkedHashSet<>();
         AuthCache authCache = new BasicAuthCache();
-        for (HttpHost host : hosts) {
-            Objects.requireNonNull(host, "host cannot be null");
-            httpHosts.add(host);
-            authCache.put(host, new BasicScheme());
+
+        Map<HttpHost, Node> nodesByHost = new LinkedHashMap<>();
+        for (Node node : nodes) {
+            Objects.requireNonNull(node, "node cannot be null");
+            // TODO should we throw an IAE if this happens?
```

Comment:
can you clarify the comment so it is clear that it's about nodes with same host?